### PR TITLE
Operator: add Network type

### PR DESCRIPTION
**Pull request by squeed**
This is an exact copy of the NetworkConfig type from the cluster-network-operator, but renamed to be just Network.

### CODE REVIEW

**Comment by squeed** [comment]
cc @derekwaynecarr

File: operator/v1/types_network.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+// NetworkList contains a list of Network configurations
+type NetworkList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []Network `json:"items"`
+}
+
+// NetworkSpec is the top-level network configuration object.
+type NetworkSpec struct {
+	// IP address pool to use for pod IPs.
```
Comment (soltysh) [inline operator/v1/types_network.go]: clusterNetwork is a list of ...

File: operator/v1/types_network.go [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+	// Some network providers, e.g. OpenShift SDN, support multiple ClusterNetworks.
+	// Others only support one. This is equivalent to the cluster-cidr.
+	ClusterNetwork []ClusterNetworkEntry `json:"clusterNetwork"`
+
+	// The CIDRs to use for services
```
Comment (soltysh) [inline operator/v1/types_network.go]: serviceNetwork is ...

File: operator/v1/types_network.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+// Not all network providers support multiple ClusterNetworks
+type ClusterNetworkEntry struct {
+	CIDR       string `json:"cidr"`
+	HostPrefix uint32 `json:"hostPrefix"`
+}
+
+// NetworkDefinition represents a single network plugin's configuration.
+// Kind must be specified, along with exactly one "Config" that matches
+// the kind.
+type DefaultNetworkDefinition struct {
```
Comment (soltysh) [inline operator/v1/types_network.go]: NetworkDefinition

File: operator/v1/types_network.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+	OpenShiftSDNConfig *OpenShiftSDNConfig `json:"openshiftSDNConfig,omitempty"`
+
+	// OVNKubernetesConfig configures the ovn-kubernetes plugin
+	// +optional
+	OVNKubernetesConfig *OVNKubernetesConfig `json:"ovnKubernetesConfig,omitempty"`
+}
+
+// AdditionalNetworkDefinition is extra networks that are available but not
+// created by default. Instead, pods must request them by name.
+type AdditionalNetworkDefinition struct {
```
Comment (soltysh) [inline operator/v1/types_network.go]: Why this is different from the above?

File: operator/v1/types_network.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+	ServiceNetwork []string `json:"serviceNetwork"`
+
+	// The "default" network that all pods will receive
+	DefaultNetwork DefaultNetworkDefinition `json:"defaultNetwork"`
+
+	// Additional networks to make available to pods when multiple networks
+	// are enabled.
+	AdditionalNetworks []AdditionalNetworkDefinition `json:"additionalNetworks,omitempty"`
+
+	// DisableMultiNetwork specifies whether or not multiple pod network
```
Comment (soltysh) [inline operator/v1/types_network.go]: All field names are starting with lowercase, just like they appear in json.

File: operator/v1/types_network.go [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+// ClusterNetwork is a subnet from which to allocate PodIPs. A network of size
+// HostPrefix (in CIDR notation) will be allocated when nodes join the cluster.
+// Not all network providers support multiple ClusterNetworks
+type ClusterNetworkEntry struct {
+	CIDR       string `json:"cidr"`
+	HostPrefix uint32 `json:"hostPrefix"`
+}
+
+// NetworkDefinition represents a single network plugin's configuration.
+// Kind must be specified, along with exactly one "Config" that matches
```
Comment (soltysh) [inline operator/v1/types_network.go]: Type must be specified...

File: operator/v1/types_network.go (hunk 4 above)
Comment (squeed) [reply operator/v1/types_network.go]: This represents a network that can be installed by Multus, as opposed to a "primary" network.

File: operator/v1/types_network.go (hunk 1 above)
Comment (squeed) [reply operator/v1/types_network.go]: Funnily enough, I modelled this after the other Kubernetes objects, who explicitly reject the convention of including the variable name in the docblock.

File: operator/v1/types_network.go (hunk 3 above)
Comment (deads2k) [reply operator/v1/types_network.go]: > NetworkDefinition

I'm willing to give this to @squeed if he wants it.  The straight move here will allow him to easily make his switch.

**Comment by deads2k** [comment]
Doc is worth updating, but to ease the transition, I'd take an exact copy of type names.

File: operator/v1/types_network.go (hunk 3 above)
Comment (soltysh) [reply operator/v1/types_network.go]: I can live with hits, but make type and its doc consistent.

File: operator/v1/types_network.go (hunk 4 above)
Comment (soltysh) [reply operator/v1/types_network.go]: Right, but I don't see any reason why the config can't be shared. I'm talking from API pov, not networking, to make it crystal clear :smile:

**Comment by squeed** [comment]
OK, docblocks updated. PTAL.

**Comment by squeed** [comment]
Ping? @soltysh I'd like to get this merged so I can move forward with the transition.

BTW, who creates the CRD? Is that still the network operator (via the CVO)?

File: operator/v1/types_network.go (hunk 6 above)
Comment (soltysh) [reply operator/v1/types_network.go]: This is still not fixed, change this to `Type`, don't forget the one below too.

**Comment by soltysh** [comment]
> BTW, who creates the CRD? Is that still the network operator (via the CVO)?

https://github.com/openshift/cluster-config-operator
@deads2k for confirmation ^

**Comment by squeed** [comment]
@soltysh sorry for the brain fart, updated.

**Comment by squeed** [comment]
ping @soltysh @deads2k for final review.

**Comment by squeed** [comment]
ping @soltysh @deads2k for review, again. Unless this merges today, this won't get in to the last beta.

File: operator/v1/types_network.go [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+	// 'false' and multiple network support is enabled.
+	DisableMultiNetwork *bool `json:"disableMultiNetwork,omitempty"`
+
+	// deployKubeProxy specifies whether or not a standalone kube-proxy should
+	// be deployed by the operator. Some network providers include kube-proxy
+	// or similar functionality. If unset, the plugin will attempt to select
+	// the correct value, which is false when OpenShift SDN and ovn-kubernetes are
+	// used and true otherwise.
+	// +optional
+	DeployKubeProxy *bool `json:"deployKubeProxy,omitempty"`
```
Comment (soltysh) [inline operator/v1/types_network.go]: I've talked with @squeed about this being pointer and he presented reasonable arguments for that. So I'll leave it as a pointer.

**Pull request merged**